### ISSUE TITLE
Questions about data resiliency and qemu

### ISSUE THREAD

**Issue by guestisp** [lizardfs/lizardfs#534]
Some questions about datar resiliency:

1) when client writes to a chunkserver, the ACK from chunkserver to the client, is sent after ALL chunkservers have properly stored the chunk or after the first chunk is wrote ? In other words, with "goal=3", the ACK to the client is sent after all 3 copies are stored or only after the first one?

2) what happens in case of "primary" chunk server failure between replication ? Is client able to know that file isn't wrote properly ? In example:
client write to chunkserver A. Write is done properly on that chunkserver then before replication is made on chunkserver B, chunkserver A dies. What happens now ? Is client able to detect this failure ?

3) what happens in case of metalogger crash during a write ? Is client able to detect the failure ?

4) is possible to know the 'delay' between metadata master and shadows ? if master fails and shadow is not 100% replicated, this lead to data-loss as all chunks wrote before master failure but still not replicated to shadow before master failure, are lost.

5) any ETA about native module for qemu to bypass FUSE ?

**Comment by Blackpaw** [comment]
Thanks for asking, was in the middle of composing those exact same question. Very curious as to the answers as well.

**Comment by guestisp** [comment]
i think that write process is fully protected. client is acknowledge only after all chunkservers are wrote, but what about metadata server?

**Comment by guestisp** [comment]
No replies here or in mailing list about this ? I'm inclined to believe that this is a SPOF as no one is publicly replying

**Comment by Zorlin** [comment]
@guestisp Be patient, you won't always get immediate responses here, especially right near Easter!

**Comment by psarna** [comment]
1. Yes, ACK from chunkserver is sent after all chunkservers stored the chunk.
2. Client will receive information that write failed, then it will retry the operation a number of times and in the end return error if all attempts failed.
3. Metadata is broadcasted to metaloggers and shadows asynchronously, client will never receive anything about their status during a write operation.
4. The delay between master and shadows depends on a lot of things. If some chunks were created but not yet acknowledged by shadow, then they may not be visible after a failover. If chunks already existed, the data may potentially still be up to date (if all chunkservers finished the writing process and chunk version wasn't changed).
5. No ETA yet, both lizardfs-client library and qemu driver are under development.

**Comment by guestisp** [comment]
4 and 5: so these are two huge SPOF. Client is unaware of what metadata is doing, if it fails during a write already acknoledged by all chunk servers, client is already happy but files are not usable because master metadata failed and shadows are out of since. In other words: this will cause a huge dataloss. If you put a DB on Lizard, in this case you'll corrupt the whole DB as some chunks are missing and can't be restored. The same is valid for VM images. If you loose 64MB being part of a huge VM image, you'll end with VM corruption.

**Comment by guestisp** [comment]
@psarna is possible to have an update on this ?

**Comment by psarna** [comment]
Writing works like that:
1. File is open - in this stage it is either created or its access time is updated, this happens in master only.
2. If needed, chunks are created in master server (by chunks here I mean chunk descriptors, like where they are, what's their id, etc. - metadata only).
2. Client receives location of every chunk it wants to write (chunkserver addresses) from master
3. Data is written to chunkservers
4. Success is acked to master

The point is that if master fails during part 3, which takes the most time, it's almost 100% probable that all metadata changes are already forwarded to shadow. Then, if master fails, shadow has all information needed to see that the file is fine. Sometimes chunk versions will be mismatched and that's what `lizardfs filerepair` command is for - with a little admin help lot of things can be restored in case of serious mismatch.
Before a chunk is written to, it is created in master first (and propagated to shadow immediately).

So, true, it is possible that when master crashes unexpectedly, new chunks of a file will not be there, because metadata was not propagated yet. It is also true that VM images are usually big files with length that doesn't change that often, so "new chunks" is not a frequent thing at all and a time window for a master crash that would actually cause some chunks to disappear is really, really small.

tl;dr: Yes, files that are written to exactly at the time when master server crashes and metadata is not yet propagated to shadows can be corrupt, nobody can argue with that. The point is, this situation is very highly unlikely. Also, the metadata of affected files can be restored manually in, I dare say, "most" cases.

**Comment by guestisp** [comment]
How can I restore metadata manually?
Are you saying that I can create proper metadata directly from chunk files?

I have some more questions about this, I'll be back shortly. :)

**Comment by psarna** [comment]
By manually I mean `lizardfs filerepair` can actually restore files in case there's a version mismatch. And if you're really desperate to retrieve some things, you can actually investigate what's left on crashed master server and try to restore data manually by mangling with the files, but this is a kind of thing that should be done by support@lizardfs.com if anybody, as it's so easy to ruin your installation by editing the files without deep knowledge of LizardFS internals.

**Comment by guestisp** [comment]
Ok, i'm back.
Let's try to explain how Lizard works, for noobs or for non developers. I'm really, realy semplifying here.

Master servers holds metadata for each file. A structure like the following (in JSON, to be more readable by humans, consider this as pseudo-code, it doesn't have to reflect the right structure):

```
{
	"id": "12334141",
	"uid": "100",
	"gid": "100",
	"perms": "0664",
	"size": "167772160",
	"chunks": [{
			"sequence": "00000001",
			"server": "192.168.0.1",
			"size": "67108864",
			"path": "chunks00/chunk_0000000000001BB3_00000001.mfs"
		},
		{
			"sequence": "00000002",
			"server": "192.168.0.2",
			"size": "67108864",
			"path": "chunks00/chunk_0000000000001BB3_00000002.mfs"
		},
		{
			"sequence": "00000003",
			"server": "192.168.0.3",
			"size": "33554432",
			"path": "chunks00/chunk_0000000000001BB3_00000003.mfs"
		}
	]
}
```

here we have a record inside master server describing a 160MB file and it's 3 chunks (two of 64MB and 1 of 32MB) pointing to the right servers (let's ignore the replication)

This is the only things that needs to be replicated to the shadow. It should be pretty fast to replicate, it's just a couple of KB.

Some questions:
1) master preallocate files on chunks, by creating empty files that would be overwritten with data by client ?

2) client starts to write chunks only after master server succefully acknowledge the creation of metadata record on it's own RAM ?

Now, based on your response, what happens if:

a) master fails when client is writing to the chunk server?
b) master fails after client has successfully wrote on the first chunkserver but chunkserver failed to write their replica ?
c) let's assume a 100GB VM. master preallocate all chunks up to 100GB? If yes, file changed inside VM doesn't trigger new chunks creation, only changes, because 100GB is already pre-allocated. What happens if master loose sync with the shadow ? Nothing because metadata for these chunks won't be changed by writing inside the VM?


tl;dr

this SPOF is possible only when:

metadata are changed AND chunks are changed AND shadow if out of sync AND (master fails before shadow is synced AND after chunks are wrote/updated by client), right? 

In which cases metadata are changed ?
a) every time a change is made to the file. (change permission, change owner, shrink the VM image, grow the VM file, ....)

b) every time file is accessed (for updating access time)

c) every time "someone" change content from inside the VM (in example: a Linux VM stored on LizardFS, from the VM i'll run `echo "hello world" > /tmp/foo`

**Comment by psarna** [comment]
~~~
master preallocate files on chunks, by creating empty files that would be overwritten with data by client ?
~~~
Yes, metadata is created first. Clients can write only to existing chunks, so the process of writing actual data will start only after all chunks are created, i.e. master knows where they are and so do appropriate chunkservers.

~~~
client starts to write chunks only after master server succefully acknowledge the creation of metadata record on it's own RAM ?
~~~
Yes, not only that, but if new chunks need to be created, chunkservers have the empty chunks prepared as well.
~~~
Now, based on your response, what happens if:

a) master fails when client is writing to the chunk server?
b) master fails after client has successfully wrote on the first chunkserver but chunkserver failed to write their replica ?
c) let's assume a 100GB VM. master preallocate all chunks up to 100GB? If yes, file changed inside VM doesn't trigger new chunks creation, only changes, because 100GB is already pre-allocated. What happens if master loose sync with the shadow ? Nothing because metadata for these chunks won't be changed by writing inside the VM?
~~~
a) Not much. Chunks are locked in master during a write operation and unlocked after write succeeded. If master does not get any ack (because it crashed), some chunks might be locked for a while longer. They will be unlocked after a timeout though.
b) _this one is complicated and has many potential cases, so I'll answer later (tomorrow I guess), because now I'm kinda busy. If for some reason I forget about it, please bump this thread_
c) What could happen is some chunk version mismatches, but if shadow knows each chunk location anyway, then no chunks will disappear from the system.

**Comment by guestisp** [comment]
> Not much. Chunks are locked in master during a write operation and unlocked after write succeeded. If master does not get any ack (because it crashed), some chunks might be locked for a while longer. They will be unlocked after a timeout though.

Ok, but is client aware of this ? If master is crashed AND client has still not finished to write chunks, client get a write error somewhere ?

> What could happen is some chunk version mismatches, but if shadow knows each chunk location anyway, then no chunks will disappear from the system.

But in this case VM is still runnng and working properly ?

> the SPOF is: metadata are changed AND chunks are changed AND shadow if out of sync AND (master fails before shadow is synced AND after chunks are wrote/updated by client), right?

is this correct ?

**Comment by psarna** [comment]
Ok, next batch of answers:
b) Generally, if any chunkserver fails to write data or distribute it to other chunkservers, client will, of course, get a write error. Then, client will retry the operation for a configurable number of times.

Yes, if master fails during a write operation (i.e. somewhere during sending data to chunkservers), client will receive an error afterwards, because it will be unable to ack write operation to master.

Generally, VM will still be running and working properly. If a lot of things crash at the same time and you end up with chunks with outdated versions only (which could happen if some chunkserver crashed as well), then things like this may need manual help, which is `lizardfs filerepair` utility.

If metadata is changed AND chunks are added AND shadow didn't manage to receive all up-to-date metadata AND master failed before it was able to send it, then files which were in the middle of the writing process MAY be corrupt. It's a lot of "AND"s to still call it a SINGLE-point-of-failure though ;) And, the files might still be in some kind of recoverable state.

**Comment by guestisp** [comment]
> If metadata is changed AND chunks are added AND shadow didn't manage to receive all up-to-date metadata AND master failed before it was able to send it, then files which were in the middle of the writing process MAY be corrupt. It's a lot of "AND"s to still call it a SINGLE-point-of-failure though ;) And, the files might still be in some kind of recoverable state.

A simple network failure (yes, it could be redundant) or a high load on shadow could cause all of this at once. Master is working, shadow is out of sync due to high load, then master fails. From this situation you can't recover, because your shadow is out of date and chunks stored on disks was updated (and acknowledge by master, but not synced to shadow)

tl;dr: the whole system resiliency is based on shadow kept in sync. If shadow was out of sync for whatever reason, you can't recover. I think that would be good to add a shadow ack (as configurable option). If you have 3 shadows, you have to get at least one ack from one of them to allow writes. This will solve all of this nasty things at once.

Anyway: in which cases metadata are updated? In example: If I create a huge loop file (let's say 100GB), mount that somewhere and then start to write files *inside* the loop file, this will trigger metadata updates ?

Something like this:

```
$ fallocate -l 209715200 /tmp/test.img
$ mkfs.ext4 /tmp/test.img 
$ sudo mount -o loop /tmp/test.img /mnt/
$ echo 'hello world' | sudo tee /mnt/testfile
hello world
```

the `echo 'hello world' | sudo tee /mnt/testfile` that changes content inside the huge file, will trigger some metadata updates in LizardFS ?

**Comment by guestisp** [comment]
I've tried the above.
Changing any file inside a disk image stored on Lizard, trigger some metadata updates.

Why this happens ? The *content* of the image should be transparent to Lizard and not triggering any metadata updates.

**Comment by guestisp** [comment]
Probably because the image is changing it's "change" time.

**Comment by guestisp** [comment]
Any update about my latest questions ?

**Comment by guestisp** [comment]
Bump

**Comment by guestisp** [comment]
Bump. Is possible to have an answer?
Why changing content inside a huge image file, also change the image file metadata?

**Comment by psarna** [comment]
Ah, sure. Changing file data can trigger:
- atime update - because the file was accessed (usually this update is lazy and doesn't happen on EVERY access, but can still occur)
- mtime update - because file's data was modified

So, ctime (metadata change time) is pretty much the only POSIX timestamp that would remain intact.

**Comment by guestisp** [comment]
So, If I set 'noatime' on lizard mount, the atime should not change, right ?
Is also possible to skip the mtime update? 

I would like to totally remove (or reduce as much as possible) the metadata interaction when using VM images stored on Lizard. Reducing these changes, will also reduce the probability of a metadata failure not synced to any shadows.

Anyway, last time i've checked, a simple change inside the image file, will trigger hundreds of metadata changes. I don't think this is normal.

**Comment by Blackpaw** [comment]
> Anyway, last time i've checked, a simple change inside the image file, will trigger hundreds of metadata changes

How do you check that?

**Comment by guestisp** [comment]
By making changes inside the image file and then looking at Lizard web interface. metadata version will grow up quickly

**Comment by psarna** [comment]
There is no "turn off mtime" option in LizardFS right now, and I don't think Linux kernel offers one either.

What do you mean hundreds of changes? What kind of changes? Modifying 1 file should trigger 1 mtime update.

@Blackpaw 
You can peek some metadata changes by reading /var/lib/mfs/changelog.mfs* files. Don't edit them though!

**Comment by guestisp** [comment]
This is what I have now:
```
$ mount | egrep 'prova|lizard'
mfs#mfsmaster:9421 on /mnt/lizard type fuse (rw,nosuid,nodev,noatime,user_id=0,group_id=0,default_permissions,allow_other)
/mnt/lizard/test.img on /mnt/prova type ext4 (rw,relatime,data=ordered)
```

`/mnt/lizard/test.img` is a loop file stored on Lizard and mounted on `/mnt/prova`

Current metadata version is 161

```
cd /mnt/prova
$ echo test | sudo tee hello
[sudo] password di ale: 
test
```

now metadata version is 189 and still growing

**Comment by Blackpaw** [comment]
But your not just writing "test" to a lizardfs file system, you're writing test to a ext4 file system hosted on lizardfs, that will be updating the meta data and block allocations of the ext4 system which will be flushed at its leisure to the underlying lizardfs system.

Its seems an overly complicated way to test the meta data writes with portions of it happening without your knowledge or control. Why not just write direct to the lizardfs system?

cd /mnt/lizardfs
$ echo test | sudo tee hello

**Comment by guestisp** [comment]
Because i'll use some VM stored on Lizard, i'll not write directly to lizard.

And I would like to skip metadata update (as much as possible) when some changes are made *inside* the VM images.

The image file is not thinprovisoned, thus any changes inside that should be transparent for Lizard and just trigger a chunk change.

All of this would be non-sense when lizard provide a way to keep master and shadow in sync with consistency check. Currenty, if "something" happens between master and shadow and master goes down, data are lost or get corrupted. I'm trying to fix this.

**Comment by guestisp** [comment]
Maybe by using a qcow2 image would be different than using a loop file mounted with an ext FS ?

**Comment by Blackpaw** [comment]
>The image file is not thinprovisoned, thus any changes inside that should be transparent for Lizard and just trigger a chunk change.

I see your point in that regard and I doubt a qcow2 file would be any different, they are both images containing file systems.

Not sure what is tracked by meta data - will updating the contents of a file without changing its size etc update metadata apart from the timestamps? question for Piotr.

**Comment by 4Dolio** [comment]
I'm not sure why there is a desire to minimize the metadata change, it is the equivilant the the superblock or inode table part of other filesystems, and so changes occur there all the time.  The shadow mechanics are pretty robust and real time.  I also store data in loopback containers and also in sparse iSCSI fileIO within LizardFS and those work great for me.  I have never bothered to watch the rate of the metadata change, but here is a look at my masters for my LFS backed iSCSI target masters that my XenServer is using as SRs:

Metadata Servers
1	lfs-us-isr1-01	shadow	connected	437794434
2	lfs-us-isr1-02	master	running	437794434
3	lfs-us-isr1-09	shadow	connected	437794434
4	lfs-us-isr1-10	\shadow	connected	437794434

The shadows are never out of sync.  I use corosync+pacemaker to manage master faults.  But more often then faults I use it to transition between masters and the VMs are never the wiser and don't skip a beat.  They do sometime hang for second or two during a master transition but otherwise do not seem to mind at all.  So I am not sure why having a metadata version of 437.7 million (dozen VMs running for 6 months) or how quickly the metadata increases would be something to worry about...

**Comment by 4Dolio** [comment]
I think that the LFS metadata is going to see all of the activity that the internally hosted filesystem is doing.  So for ext4:
* journal intends to write a file (chunk holding journal updated)
* file data written (chunk holding file data updated)
* inod update for file data (chunk holding superblock changed)
* journal confirms finished op (chunk holding journal updated)
Note that the journal and superblocks are stored multiple times for a filesystem, so each of those will cause changes for each copy of the superblock/journal.

Perhaps use ext3 or ext2 for the loopback filesystem as it lacks the journal?
I have used ZFS and EXT4 within lookbacks and within iSCSI sparse fileIO. they both work pretty good.

**Comment by guestisp** [comment]
Yes, this is true, but currently there is a huge SPOF in lizard.
Shadow and master doesn't acknowledge each other, if, for whatever reason, shadow get out of sync, master would be unaware of this and still allow changes to the cluster.

Now in case of master daily, you'll end up with a corrupted cluster because the shadow where are you restoring from is out of sync.

Shadows could go out of sync for multiple reasons, like a network congestion , high I/o or similar
In this case your cluster is NOT high available anymore and would be better to bring down the whole cluster to preserve data consistency

In other words, you have a huge splitbrain

**Comment by 4Dolio** [comment]
Yes. But that has never occurred in the 2 years of LizardFS and an additional 3 years of MooseFS.  The LizardFS Shadow is basically MooseFS-MetaLogger that loads the metadata stream live into a master process which can take over at a moments notice.  The shadow process includes data integrity overhead so a shadow would know if it is not properly in sync.  The only time that has happened to me is during shadow start due to a hardcoded time limit on how quickly the get-into-sync process is allowed to take.  From that experience it seems that a shadow which is not in sync including metadata checksum integrity checks then it will retry as quickly and often as the master will allow, which is tunable.  But as I opened with, I have never encountered an instance where a running shadow has errored out of sync once in sync.  There is no conceivable way that could happen I think.  And if it could somehow the cluster would know and would correct for it without intervention.

I acknowledge the master might not get positive feedback from the shadows regarding their data consistency, but it need not care about that.  The mfs.cgi web interface I believe represents the master polling the shadows directly, or at least identifying the the web cgi who the shadows are so it may poll the shadows for their metadata revision.  It is all on the shadows to never accept corruption of the metadata stream.  I am pretty certain that as I described earlier it is already doing that but should never normally need to.

A network outage would not cause a shadow to go out of sync.... assuming the network returns to a working state and the TCP layer had done it's job to deliver the missing metadata packets which got held up temporarily.  Should TCP fail to deliver the shadow would notice and upon the addition of faulty metadata the shadow would request a full resync and resolve itself.  I would think it would be difficult to deploy a network which would impact the master shadow(s) communications like this without just broadly being an unusable network.

Neither masters nor shadows are dependent on disk IO beyond their starting, stopping, and periodic dumps.  In fact, I believe you could run a master/shadow out of disk to the point that those start/stop/dump operations can fail, of course potentially making that particular master/shadow useless for a cold start from disk.  But the metadata master or shadow functions in RAM would continue to work.  I believe one could even start a shadow service with no disk at all. In theory you could run all masters with no disk and perpetually transition from master to shadow assuming you never ever never want to stop and later start your storage..

There is even a mechanic such that a master with no shadows will refuse to uncleanly exit without flushing the metadata to disk safely as it can. A normally operating cluster does not use the disk copies of the metadata except for to cold stop and cold start.  It is a purely ram+network+handoff process while transitioning between masters and shadows for normal maintenance processes and faults alike.  By faults I mean with prerequisites such as network or system or even master crash faults.  The shadow does have it's own metadata integrity features which I believe you can turn off with a tunable to ignore the metadata checksums and just blindly incorporate garbage.

It would not be better to bring down the whole cluster because shadows have no role in the working of the cluster, except for metadata authority transitions and failovers.  Running one shadow is the bare minimum for that feature but none are required for the normal operation of the cluster.  One shadow has been enough for my production storage for the past two years.  I do have more shadows for some more important LizardFS clusters we use.  All our chunkservers also act as pre-shadow metaloggers so there is yet another copy of the one authoritative metadata stream.  The master knows the difference between loggers and shadows.  The master need not worry about the state of loggers nor shadows because they are not required for a functioning cluster. 

You can not have split brain nor other cluster related fault situations if you do not cause them yourself by promoting/starting more than one authoritative master at a time.  That particular problem is not something that a storage system needs to solve. Master HA is not problem for LizardFS to solve because such HA orchestration is a complex problem in and of itself.  There are numerous solutions available for the HA problem and you should take into consideration when choosing pros and cons such as dealing with split brain.

Let me amend that last topic, LizardFS need not address metadata HA until such time as it adds features for multi-master authority within a single metadata set.  But that is another feature request for the far future if ever.

Does that help?  You are sort of complaining to the hard disk guys (LizardFS) that your raid controller (Master High Availability Solution You Choose) might give you problems.  HA Master automated failover is not something LizardFS does directly.  Look for the lizardfs-admin command to see the transition related calls that your HA solution can interact with.

**Comment by guestisp** [comment]
That is true, but is still a SPOF
hard to happen is different from impossible

Also a total server failure (with enterprise grade servers) is rare, but HA solution are still developed as protection.

**Comment by 4Dolio** [comment]
Yes the single metadata master is a SPOF.  Thus the need to wrap them up into a HA solution for masters and shadows.

**Comment by guestisp** [comment]
But async writes wouldn't solve the issue.
A real HA solution would be with sync writes where you'll be 100% sure that what is replicated to the shadow is the same you have on the master in any time

**Comment by asbai** [comment]
I've some questions about this thread:

1. The replication between master and shadow is synchronized, right? Is there a way to change it to a full synchronized form? For example: use some external sync tool like DRBD to replicate the master's disk to a salve node, and auto start the daemon on the slave node when the master is down? 

So, the real question is: **Will the master flush related updates to the disk before it return the ACK to client?**

2. like @psarna said: "Yes, ACK from chunkserver is sent after all chunkservers stored the chunk.". If I've a goal=3, and one of the chunck server node is down (by a network partition or hardware failure, e.g.: there is to replica can be reachable at the time), then what will happen when the try to read or write this chunk?

What will happen if only one of the three replica can be reached? Is there a quorum algorithm to ensure only the majority part can continue access the data when a network partition occured?

3. Like the questen 2, what if I'm using EC replica? For example: If I'm using a 8+4 EC scheme, then:
3.1. chould I let the fragments uniformly distributed to three different data center (4+4+4), so data still accessible even if any one DC fails?
3.2. Like the questen 2, Will the EC replica using a quorum algorithm to achieve the strong consistency and partition tolerance with care of the high availability at the same time?

**Comment by asbai** [comment]
Another question：I see the client will read data from the nearest Chunk server, is the cluster guaranteed the read consistency (May a read request retrieves obsolete data) ?

Is there a way to guarantee the strong consistency for the simple N's replica and the EC replica?

**Comment by asbai** [comment]
No answer here ?

**Comment by cloudweavers** [comment]
When a write is performed, the write acknowledge is received only when the number of chunkserver that acknowledge the write is equal to the goal. So, if you have goal 2, two chunkserver must report a write. This means that reading from the nearest chunk server (if that chunk server has the requested chunk) will be always coherent.
